### ISSUE TITLE
twisted.names.dns.DNSMixin._query should return a cancellable Deferred.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6644]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kaizhang reported|
|-|-|
|Trac ID|trac#6644|
|Type|enhancement|
|Created|2013-07-27 08:17:20Z|
|Branch|https://github.com/twisted/twisted/tree/dnsmixin-deferred-cancellation-6644-2|

`twisted.names.dns.DNSMixin._query` should return a cancellable `Deferred`. When cancelling the `Deferred`, `resultDeferred` and `cancelCall` should be removed from `DNSMixin.liveMessages`. And the `cancelCall` should be cancelled.

<details><summary>Searchable metadata</summary>

```
trac-id__6644 6644
type__enhancement enhancement
reporter__kaizhang kaizhang
priority__normal normal
milestone__ 
branch__branches_dnsmixin_deferred_cancellation_6644_2 branches/dnsmixin-deferred-cancellation-6644-2
branch_author__kaizhang kaizhang
status__new new
resolution__None None
component__names names
keywords__ 
time__1374913040000000 1374913040000000
changetime__1382195453000000 1382195453000000
version__None None
owner__kaizhang kaizhang

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set owner to kaizhang<br>kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|kaizhang commented|
|-|-|

(In [39315]) Branching to 'dnsmixin-deferred-cancellation-6644'

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dnsmixin-deferred-cancellation-6644)

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|kaizhang set owner to kaizhang|
|-|-|

Need to add news file.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|kaizhang removed owner|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dnsmixin-deferred-cancellation-6644)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

You should have a test for the case where a response does eventually arrive after cancellation, making sure nothing breaks and it's just dropped with no exceptions.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|kaizhang commented|
|-|-|

Replying to [itamar](#note_6):
> You should have a test for the case where a response does eventually arrive after cancellation, making sure nothing breaks and it's just dropped with no exceptions.

I added the tests for this case in another ticket since the arrived response is handle by `DNSProtocol`/`DNSDatagramProtocol`. The ticket is [#6655](https://github.com/twisted/twisted/issues/6655). When I wrote the tests I found a bug about `twisted.names.test.test_dns.TestTCPController` so I fixed the bug and added the tests in a new ticket.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj commented|
|-|-|

There are some questions in ticket:6659#comment:2 about whether the timeout parameter provided to queryTCP and lookupZone should be applied overall ie time taken to establish TCP connection and time taken for complete DNS response.

And if so how does it fit in with this new deferred cancellation function?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

I'll add comment about overall timeouts in [#6659](https://github.com/twisted/twisted/issues/6659). As far as Deferred cancellation, it is a nice way to implement timeouts (untested, but there's a ticket somewhere to add this to Twisted):

```
def timeout(deferred, seconds):
    delayedCall = reactor.callLater(seconds, deferred.cancel)
    def checkCanceled(result):
        if delayedCall.active(): delayedCall.cancel()
        return result
    deferred.addBoth(checkCanceled)
```

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|kaizhang set owner to kaizhang<br>kaizhang set status to `assigned`|
|-|-|

Need to rebranch to get the tests for response arriving after cancellation from [#6655](https://github.com/twisted/twisted/issues/6655).

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|kaizhang commented|
|-|-|

(In [39658]) Branching to 'dnsmixin-deferred-cancellation-6644-2'

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dnsmixin-deferred-cancellation-6644-2)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince commented|
|-|-|

A couple of comments:
 1. `test_cancelQueryCancelTheCancelCall`, you should cancel the deferred returned by `_query` rather than the one stored in `liveMessages`. They are the same, but the later is an implementation detail.
 2. If a request is cancelled, subsequent responses with that id are passed to the DNS controller. This is also what happens whenthe timeout is reached. I think it would make more sense to silently drop them. `twisted.names.client.Resolver` will log any messages that it receives (although, it ends up not being an issue in this case, since the protocol is closed immediately when the query is cancelled).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj set owner to kaizhang<br>@wallrj set status to `new`|
|-|-|

Thanks kaizhang,

Sorry for the delay in actually reviewing this ticket.

Notes:
 * Merges cleanly
 * Tests pass locally
 * Coverage is complete
 * Builds all green (apart from a couple of spurious twistedchecker warnings)

In addition to Tom's numbered points above:
 1. twisted/names/dns.py
    1. I don't like the fact that the nested function refers to variable that is
       defined after the function. I know it works, but somehow it doesn't look
       right to me. I also not that you've used this technique in previous
       cancellation branches which have been accepted and merged.
    2. So consider making cancel a method instead of a nested function. eg
       ```
       #!diff
       Index: twisted/names/dns.py
       ===================================================================
       --- twisted/names/dns.py	(revision 40401)
       +++ twisted/names/dns.py	(working copy)
       @@ -2150,23 +2150,29 @@
       except:
       return defer.fail()

       - def cancel(deferred):
       - """
       - Cancel the query.
       -
       - Remove the C{resultDeferred} and C{cancelCall} from
       - L{self.liveMessages}. Cancel the C{cancelCall}.
       -
       - @param deferred: The cancelled L{defer.Deferred}.
       - """
       - del self.liveMessages[id]
       - cancelCall.cancel()
       - resultDeferred = defer.Deferred(cancel)
       - resultDeferred = defer.Deferred(self._queryCancel)
         cancelCall = self.callLater(timeout, self._clearFailed, resultDeferred, id)
         self.liveMessages[id] = (resultDeferred, cancelCall)

         return resultDeferred

       -
       - def _queryCancel(self, deferred):
       - """
       - Cancel the query.
       -
       - Remove the C{resultDeferred} and C{cancelCall} from
       - L{self.liveMessages}. Cancel the C{cancelCall}.
       -
       - @param deferred: The cancelled L{defer.Deferred}.
       - """
       - for id, (resultDeferred, cancelCall) in self.liveMessages.items():
       - if resultDeferred is deferred:
       - del self.liveMessages[id]
       - cancelCall.cancel()
       - break
       -
       -
                  def _clearFailed(self, deferred, id):
                      """
                      Clean the Deferred after a timeout.
       ```
    3. I also wonder if this is slightly safer. Imagine if someone called
       DNSDatagramProtocol.query with a duplicate ID, the original entry in
       liveMessages would be overwritten and a cancellation of the original
       query would then delete the wrong liveMessages entry.
    4. The *cancel* docstring contains *L{self.liveMessages}* which I don't
       think is valid. L{} has to refer to an object that can be found by
       PyDoctor.
    5. I've been thinking about itamar's proposed timeout function which calls
       cancel. It sounds neat, but what about if I need to distinguish between a
       cancellation and a timeout. A Timeout would probably result in a
       scheduled retry whereas cancellation would not. Is there anyway to
       customise the type of Failure in the errback as a result of cancellation?
    6. Should cancellation also close the underlying transport? I was
       considering what happens if a result eventually comes back from a
       cancelled query. It will be treated as a new message and dispatched to
       controller.messageReceived.
       1. Tom makes the same point in comment:13 -- he suggests silently
          dropping the message, which would require some extra logic in data and
          datagram received....I think (plus associated tests)
       2. I wonder if cancellation should just terminate the connection (in the
          case of TCP) or close the port (in the case of UDP)?
       3. ...but then again, what if the user wants to issue multiple queries
          using the same DNSProtocol instance? I can't think why, since the
          reason we've cancelled the query is presumably that the server that
          we're talking to has not responded, so why bother sending further
          queries or expecting results to other outstanding queries to that
          server.
 2. twisted/names/test/test_dns.py
    1. It doesn't seem right to be testing only the DNSMixin class.
       1. Wouldn't it be better to test the DNSDatagramProtocol and DNSProtocol
          instead (or in addition).
       2. I think that would make it easier to refactor the public DNS protocols
          in the future. (ie if we wanted to replace or remove the
          DNSMixin...which I'd quite like to do because it confuses me).
       3. In any case, I think it might be clearer to call the test case
          CancellationTests or CancellationTestCase.
       4. Say "is" instead of "should be" in the new test docstrings.

 3. twisted/names/topfiles/6644.feature
    1. The news file should refer to the affected public APIs, not DNSMixin._query.


Here are some situations where I think query cancellation will be useful.
 * client.Resolver -- cancelling an unanswered UDP query before issuing a TCP
   query.
 * EDNSProtocol -- cancelling an unanswered EDNS query before issuing another query with
   a smaller UDP packet size.
 * root.Resolver -- issuing parallel queries to all authoritative nameservers
   and cancelling outstanding queries after the first response is received.
 * DNSSEC verification -- again, I imagine we will issue parallel queries to
   auth servers and lock onto the server which responds fastest, cancelling
   other outstanding queries.
 * GAI endpoint -- where IPv4 and IPv6 UDP DNS queries are issued in parallel,
   cancel the oustanding query when the first response is received.

Anyway, thanks again. Please address or answer the numbered points above
(including Tom's) and resubmit for another review.

-RichardW.
